core/domain/classifier_services.py

```diff
@@ -0,0 +1,507 @@
+        - 'prediction_confidence': the prediction confidence. This is
+        Prob(the doc should be assigned this label |
+        the doc is not assigned _DEFAULT_LABEL).
+        - 'all_predictions': a dict mapping each label to
+        Prob(the doc should be assigned to this label).
+
+        Because _DEFAULT_LABEL is a special label that captures unspecial
+        tokens (how ironic), its probability is not a good predicting
+        indicator. For the current normalization process, it has on
+        average a higher probability than other labels. To combat this,
+        all other labels are normalized as if the default label did not
+        exist. For example, if we have two labels with probability 0.2
+        and 0.3 with the default label having probability 0.5, the
+        normalized probability for the two labels without the default
+        label is 0.4 and 0.6, respectively.
+
+        The prediction threshold is currently defined at the classifier level.
+        A higher prediction threshold indicates that the predictor needs
+        more confidence prior to making a prediction, otherwise it will
+        predict _DEFAULT_LABEL. This will make non-default predictions more
+        accurate, but result in fewer of them."""
+        default_label_id = self._get_label_id(self._DEFAULT_LABEL)
+        prediction_label_id = default_label_id
+        prediction_confidence = 0
+        label_probabilities = self._get_label_probabilities(d)
+        normalization_coeff = (
+            1.0 / (1.0 - label_probabilities[default_label_id]))
+
+        for l, prob in enumerate(label_probabilities):
+            if (l != default_label_id and
+                    prob * normalization_coeff > self._prediction_threshold and
+                    prob * normalization_coeff > prediction_confidence):
+                prediction_label_id = l
+                prediction_confidence = prob * normalization_coeff
+
+        return {
+            'prediction_label_id': prediction_label_id,
+            'prediction_label_name':
+                self._get_label_name(prediction_label_id),
+            'prediction_confidence': prediction_confidence,
+            'all_predictions': label_probabilities
+        }
+
+    def _parse_examples(self, examples):
+        """Unzips docs and label lists from examples and returns the two lists.
+
+        Docs are split on whitespace. Order is preserved."""
+        docs = []
+        labels_list = []
+        for example in examples:
+            doc_string = example[0]
+            doc = doc_string.split()
+            if len(doc) > 0:
+                labels = example[1]
+                docs.append(doc)
+                labels_list.append(labels)
+        return docs, labels_list
+
+    def _iterate_gibbs_sampling(self, iterations, doc_ids=None):
+        """Runs Gibbs sampling for "iterations" number of times on the provided
+
+        docs."""
```

Comment:
or, better, actually, just pass in xrange(self._num_docs) from the caller in that case, and remove the default arg